Cosine.lua

```diff
@@ -0,0 +1,163 @@
+local Cosine, parent = torch.class('nn.Cosine', 'nn.Module')
+
+function Cosine:__init(inputSize,outputSize)
+   parent.__init(self)
+
+   self.weight = torch.Tensor(outputSize,inputSize)
+   self.gradWeight = torch.Tensor(outputSize,inputSize)
+   
+   self:reset()
+end
+
+function Cosine:reset(stdv)
+   if stdv then
+      stdv = stdv * math.sqrt(3)
+   else
+      stdv = 1./math.sqrt(self.weight:size(1))
+   end
+   self.weight:uniform(-stdv, stdv)
+end
+
+function Cosine:updateOutput(input)
+   local inputSize = self.weight:size(2)
+   local outputSize = self.weight:size(1)
+   
+   self._weightNorm = self._weightNorm or self.weight.new()
+   self._inputNorm = self._inputNorm or self.weight.new()
+   
+   -- y_j = (w_j * x) / ( || w_j || * || x || )
+   
+   self._weightNorm:norm(self.weight,2,2):add(1e-12)
+   if input:dim() == 1 then
+      self.output:resize(outputSize):zero()
+      self.output:addmv(1, self.weight, input)
+      self.__norm = input:norm()+1e-12
```

Comment:
@fmassa Good question. Personally, I like that the 1D case is different than 2D, as it makes for a better unit test. In anycase, if you want your 1D use case to be fast, you could just make the 1D input into a 2D Tensor *before* it enters this module. But yeah, in any case, the code is redundant. If you want to make the change, feel free, but please transfer the 1D code into the unit test though (without all the buffering of course).